src/invoke.rs

```diff
@@ -30,10 +41,99 @@ pub enum Error {
     Host(#[from] HostError),
 }
 
+pub struct Snap {
+    storage: OrdMap<LedgerKey, LedgerEntry>,
+}
+
+impl SnapshotSource for Snap {
+    fn get(&self, key: &LedgerKey) -> Result<LedgerEntry, HostError> {
+        match self.storage.get(key) {
+            Some(v) => Ok(v.clone()),
+            None => Err(HostError::General("Missing entry")),
+        }
+    }
+    fn has(&self, key: &LedgerKey) -> Result<bool, HostError> {
+        Ok(self.storage.contains_key(key))
+    }
+}
+
+pub fn read_storage(
+    input_file: &std::path::PathBuf,
+) -> Result<OrdMap<LedgerKey, LedgerEntry>, Error> {
+    let mut res = OrdMap::new();
+
+    let file = match File::open(input_file) {
+        Ok(f) => f,
+        Err(e) => {
+            //File doesn't exist, so treat this as an empty database and the file will be created later
+            if e.kind() == std::io::ErrorKind::NotFound {
+                return Ok(res);
+            } else {
+                return Err(Error::Io(e));
+            }
+        }
+    };
+
+    let mut lines = io::BufReader::new(file).lines();
+    while let (Some(line1), Some(line2)) = (lines.next(), lines.next()) {
+        let lk = LedgerKey::read_xdr(&mut line1?.as_bytes())?;
+        let le = LedgerEntry::read_xdr(&mut line2?.as_bytes())?;
+        res.insert(lk, le);
+    }
+
+    Ok(res)
+}
+
+pub fn commit_storage(
+    startup_state: OrdMap<LedgerKey, LedgerEntry>,
+    storage_map: &OrdMap<LedgerKey, Option<LedgerEntry>>,
+    output_file: &std::path::PathBuf,
+) -> Result<(), Error> {
+    //Need to start off with the startup_state since it's possible the storage_map did not touch every existing entry
+    let mut new_state = startup_state.clone();
+
+    let mut file = File::create(output_file)?;
+    for (lk, ole) in storage_map {
+        if let Some(le) = ole {
+            new_state.insert(lk.clone(), le.clone());
+        } else {
+            new_state.remove(lk);
+        }
+    }
+
+    for (lk, le) in new_state {
+        lk.write_xdr(&mut file)?;
+        writeln!(&mut file)?;
+        le.write_xdr(&mut file)?;
+        writeln!(&mut file)?;
+    }
+
+    Ok(())
+}
+
 impl Invoke {
     pub fn run(&self) -> Result<(), Error> {
         let contents = fs::read(&self.file).unwrap();
-        let h = Host::default();
+
+        // Initialize storage and host
+        // db_file format is one xdr object per line, where a LedgerKey is followed by the corresponding LedgerEntry.
+        //
+        // LedgerKey1
+        // LedgerEntry1
+        // LedgerKey2
+        // LedgerEntry2
+        // ...
+
+        // TODO: allow option to separate input and output file
+        let startup_state = read_storage(&self.db_file)?;
```

Comment:
Since this is a collection of ledger entries, suggest we name it so here:
```suggestion
        let ledger_entries = read_storage(&self.db_file)?;
```